destination_middleware.go

```diff
@@ -498,3 +505,239 @@ func (d *destinationWithRecordFormat) Write(ctx context.Context, recs []opencdc.
+		Destination:         impl,
+		defaults:            s.Config,
+		subjectVersionCache: make(map[string]map[int]schema.Schema),
+	}
+}
+
+// destinationWithSchemaExtraction is the actual middleware implementation.
+type destinationWithSchemaExtraction struct {
+	Destination
+	defaults DestinationWithSchemaExtractionConfig
+
+	payloadEnabled bool
+	keyEnabled     bool
+
+	subjectVersionCache map[string]map[int]schema.Schema
+	payloadWarnOnce     sync.Once
+	keyWarnOnce         sync.Once
+}
+
+func (d *destinationWithSchemaExtraction) Parameters() config.Parameters {
+	return mergeParameters(d.Destination.Parameters(), d.defaults.parameters())
+}
+
+func (d *destinationWithSchemaExtraction) Configure(ctx context.Context, config config.Config) error {
+	err := d.Destination.Configure(ctx, config)
+	if err != nil {
+		return err
+	}
+
+	d.keyEnabled = *d.defaults.KeyEnabled
+	if val, ok := config[configDestinationWithSchemaExtractionKeyEnabled]; ok {
+		d.keyEnabled, err = strconv.ParseBool(val)
+		if err != nil {
+			return fmt.Errorf("invalid %s: failed to parse boolean: %w", configDestinationWithSchemaExtractionKeyEnabled, err)
+		}
+	}
+
+	d.payloadEnabled = *d.defaults.PayloadEnabled
+	if val, ok := config[configDestinationWithSchemaExtractionPayloadEnabled]; ok {
+		d.payloadEnabled, err = strconv.ParseBool(val)
+		if err != nil {
+			return fmt.Errorf("invalid %s: failed to parse boolean: %w", configDestinationWithSchemaExtractionPayloadEnabled, err)
+		}
+	}
+
+	return nil
+}
+
+func (d *destinationWithSchemaExtraction) Write(ctx context.Context, records []opencdc.Record) (int, error) {
+	if d.keyEnabled {
+		for i := range records {
+			if err := d.decodeKey(ctx, &records[i]); err != nil {
+				return 0, err
+			}
+		}
+	}
+	if d.payloadEnabled {
+		for i := range records {
+			if err := d.decodePayload(ctx, &records[i]); err != nil {
+				return 0, err
+			}
+		}
+	}
+
+	return d.Destination.Write(ctx, records)
+}
+
+func (d *destinationWithSchemaExtraction) decodeKey(ctx context.Context, rec *opencdc.Record) error {
+	subject, errSubject := rec.Metadata.GetKeySchemaSubject()
+	version, errVersion := rec.Metadata.GetKeySchemaVersion()
+	switch {
+	case errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound) &&
+		errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		// log warning once, to avoid spamming the logs
+		d.keyWarnOnce.Do(func() {
+			Logger(ctx).Warn().Msgf(`record does not have an attached schema for the key, consider disabling the destination schema key decoding using "%s: false"`, configDestinationWithSchemaExtractionKeyEnabled)
+		})
+		return nil
+	case errSubject != nil && !errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get key schema subject from metadata: %w", errSubject)
+	case errVersion != nil && !errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get key schema version from metadata: %w", errVersion)
+	}
```

Comment:
WDYT about handling the case where one of the metadata fields is present and the other isn't? It should happen under normal circumstances but a bug or a strange processor might cause something like that.

Reply:
I'm not sure what you mean, as far as I can see we are handling all possible cases. We tolerate when both metadata fields are empty, but we don't accept only one being empty. Am I missing something?

Reply:
I'm thinking now if we should skip the middleware unless both are present 🤔 Then the connector can figure out what to do with that record.

Reply:
Ok I went with this: https://github.com/ConduitIO/conduit-connector-sdk/pull/155/commits/48b1d5d8d58fdc9a897c395be40438c2a432d4e8

So essentially we log the warning and skip the middleware unless the subject _and_ the version are both attached.

WDYT?

Reply:
Logging sounds good to me!